### PR TITLE
Get rid of global metric registries in AtlasDB

### PR DESCRIPTION
**Pull request by j-baker**
Breaks large internal product, but lets other internal product be a
multi client of AtlasDB without hilariously bogus metrics.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/3237)

### CODE REVIEW

**Comment by j-baker** [comment]
before merging, should also take this opportunity to nuke large internal product's other usages of weird AtlasDB constructors (which I'm breaking here).

**Comment by j-baker** [comment]
i am aware that every reference to create now says 'of', and i will fix :)

File: timelock-agent/src/main/java/com/palantir/timelock/paxos/TimeLockAgent.java [hunk 1]
```diff
@@ -153,7 +162,7 @@ private void registerExceptionMappers() {
     /**
      * Creates timestamp and lock services for the given client. It is expected that for each client there should
      * only be (up to) one active timestamp service, and one active lock service at any time.
-     * @param client Client namespace to create the services for
+     * @param client Client namespace create of the services for
```
Comment (jeremyk-91) [inline timelock-agent/src/main/java/com/palantir/timelock/paxos/TimeLockAgent.java]: nit: probably a double change here?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraVerifier.java [hunk 2]
```diff
@@ -180,7 +180,7 @@ private static boolean attemptToCreateKeyspace(CassandraKeyValueServiceConfig co
                     throw ire;
                 }
             } catch (Exception exception) {
-                log.warn("Couldn't use host {} to create keyspace."
+                log.warn("Couldn't use host {} to of keyspace."
```
[the remark this answers is not in the document]
Comment (j-baker) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraVerifier.java]: ughhhhh

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 3]
```diff
@@ -867,11 +897,14 @@ private static LockAndTimestampServices createRawRemoteServices(AtlasDbConfig co
...
+            MetricsManager metricsManager,
             Consumer<Object> env,
             com.google.common.base.Supplier<LockService> lock,
             com.google.common.base.Supplier<TimestampService> time) {
-        LockService lockService = ServiceCreator.createInstrumentedService(lock.get(), LockService.class);
-        TimestampService timeService = ServiceCreator.createInstrumentedService(time.get(), TimestampService.class);
+        LockService lockService = ServiceCreator.createInstrumentedService(
+                metricsManager.getRegistry(), lock.get(), LockService.class);
+        TimestampService timeService = ServiceCreator.createInstrumentedService(
+                metricsManager.getRegistry(),time.get(), TimestampService.class);
```
[the remark this answers is not in the document]
Comment (j-baker) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: ah :(

File: atlasdb-dagger/src/main/java/com/palantir/atlasdb/services/MetricsModule.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import dagger.Provides;
+
+@Module
+public class MetricsModule {
+
+    @Provides
+    @Singleton
+    public MetricsManager metricsManager() {
+        return MetricsManagers.createForTests();
+    }
```
[the remark this answers is not in the document]
Comment (j-baker) [reply atlasdb-dagger/src/main/java/com/palantir/atlasdb/services/MetricsModule.java]: i'll inline this. It's actually correct behaviour; they don't have metrics passed in externally.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/TargetedSweeperTest.java [hunk 5]
```diff
@@ -104,12 +105,12 @@ public void initializingTargetedSweeperWithMoreThreadsThanShardsIncreasesNumberO
...
 
         TargetedSweeper sweeperConservative = TargetedSweeper
-                .createUninitialized(null, null, DEFAULT_SHARDS + 5, 0, ImmutableList.of());
+                .createUninitialized(MetricsManagers.createForTests(), null, null, DEFAULT_SHARDS + 5, 0, ImmutableList.of());
         sweeperConservative.initialize(timestampsSupplier, spiedKvs, mock(TargetedSweepFollower.class));
         assertThat(progress.getNumberOfShards()).isEqualTo(DEFAULT_SHARDS + 5);
 
         TargetedSweeper sweeperThorough = TargetedSweeper
-                .createUninitialized(null, null, 0, DEFAULT_SHARDS + 10, ImmutableList.of());
+                .createUninitialized(MetricsManagers.createForTests(), null, null, 0, DEFAULT_SHARDS + 10, ImmutableList.of());
```
[the remark this answers is not in the document]
Comment (j-baker) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/TargetedSweeperTest.java]: oh boo

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 6]
```diff
@@ -265,20 +266,27 @@ public void testLockAfterGet() throws Exception {
...
-        SnapshotTransaction snapshot = new SnapshotTransaction(
-                kvMock,
-                new LegacyTimelockService(timestampService, lock, lockClient),
+        SnapshotTransaction snapshot = new SnapshotTransaction(metricsManager,
+                keyValueService,
+                null,
                 transactionService,
                 NoOpCleaner.INSTANCE,
+                () -> transactionTs,
+                ConflictDetectionManagers.createWithNoConflictDetection(),
```
[the remark this answers is not in the document]
Comment (j-baker) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: unintentional!

File: timelock-agent/src/main/java/com/palantir/timelock/paxos/TimeLockAgent.java (hunk 1 above)
Comment (j-baker) [reply timelock-agent/src/main/java/com/palantir/timelock/paxos/TimeLockAgent.java]: done

File: atlasdb-ete-tests/src/main/java/com/palantir/atlasdb/AtlasDbEteServer.java [hunk 7]
```diff
@@ -142,14 +151,15 @@ private TargetedSweeper initializeAndGet(TargetedSweeper sweeper, SerializableTr
...
+            TaggedMetricRegistry taggedMetricRegistry)
             throws InterruptedException {
         Stopwatch sw = Stopwatch.createStarted();
         while (sw.elapsed(TimeUnit.SECONDS) < CREATE_TRANSACTION_MANAGER_MAX_WAIT_TIME_SECS) {
             try {
-                return createTransactionManager(config, atlasDbRuntimeConfig, environment);
+                return createTransactionManager(config, atlasDbRuntimeConfig, environment, taggedMetricRegistry);
             } catch (RuntimeException e) {
-                log.warn("An error occurred while trying to create transaction manager. Retrying...", e);
+                log.warn("An error occurred while trying to of transaction manager. Retrying...", e);
```
[the remark this answers is not in the document]
Comment (j-baker) [reply atlasdb-ete-tests/src/main/java/com/palantir/atlasdb/AtlasDbEteServer.java]: done

File: atlasdb-cli/src/test/java/com/palantir/atlasdb/cli/command/TestSingleBackendCommand.java [hunk 8]
```diff
@@ -64,7 +64,7 @@ public int execute(AtlasDbServices services) {
                 services.getTransactionManager().getImmutableTimestamp();
             }
 
-            // test kvs create table
+            // test kvs of table
```
[the remark this answers is not in the document]
Comment (j-baker) [reply atlasdb-cli/src/test/java/com/palantir/atlasdb/cli/command/TestSingleBackendCommand.java]: done

File: atlasdb-dagger/src/main/java/com/palantir/atlasdb/services/MetricsModule.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply atlasdb-dagger/src/main/java/com/palantir/atlasdb/services/MetricsModule.java]: Agreed. Yep - that's why I mentioned it was fine, though it looks weird

**Comment by jeremyk-91** [comment]
... and #3188 went in which I presume is making the pile of conflicts. argh

Let me know if you want any help on fixing the targeted sweepy tests!

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/metrics/SweepMetricsAssert.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+import org.assertj.core.internal.Objects;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.collect.ImmutableMap;
+import com.palantir.atlasdb.AtlasDbMetricNames;
+import com.palantir.atlasdb.util.MetricsManager;
+import com.palantir.tritium.metrics.registry.MetricName;
+
+public final class SweepMetricsAssert extends AbstractAssert<SweepMetricsAssert, MetricsManager> {
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/metrics/SweepMetricsAssert.java]: 💯 thank you!

**Comment by jeremyk-91** [comment]
`TimestampTrackerTest` had a couple of failures :( mainly because there were some assumptions made in the old model that can be cleared out now (things like `metricsAreDeregisteredUponClose()`, for instance)

**Pull request merged**